spack.md: update instructions to use shared Spack instance (#1190)
ACCESS-Hive/access-hive.github.io: docs/getting_started/spack.md

Reviewer: There is a bug with backticks here. 
I think they need to be escaped (``/```)or HTML code `&#96;` can be used instead

!!! warning

>     **01/05/2026:** ACCESS-NRI has migrated to a shared Spack `v1.1` instance. If you previously followed these instructions to set up a personal *Spack* instance, please consider [moving the old Spack instance](remove-deprecated-spack-directories) below before using the shared Spack instance.

Reviewer: ```suggestion
    **01/05/2026:** ACCESS-NRI has migrated to a shared _Spack_ `v1.1` instance. If you previously followed these instructions to set up a personal _Spack_ instance, please consider [removing the old _Spack_ directories](#remove-deprecated-spack-directories) before using the shared _Spack_ instance.
```

# How to use Spack on Gadi for building ACCESS models

[Spack](https://spack.io/about/) is a build-from-source package manager, specifically designed to simplify the installation of scientific software on supercomputers.

To use _Spack_, please familiarise yourself with the [Basic Usage instructions](https://spack.readthedocs.io/en/latest/basic_usage.html) and [Environments](https://spack.readthedocs.io/en/latest/environments.html).

We also recommend that you refer to the [Spack 101 Tutorial](https://spack-tutorial.readthedocs.io/en/latest/).

_Spack_ allows users to build ACCESS models directly from the source code, swap model components, and carry out development testing that involves modifying the source code.

For instructions on how to build an ACCESS model using _Spack_, refer to [Modify and build an ACCESS model's source code](/models/build_a_model/build_source_code).

## Prerequisites
- **NCI Account**<br> 
    These instructions are tailored specifically for _Gadi_. To use _Spack_ on _Gadi_, you need to [Set Up your NCI Account](/getting_started/set_up_nci_account).

## Enable Spack

```
module use /g/data/vk83/modules
module load spack/1.1
```

> ## Test Spack (OPTIONAL)

Reviewer: As mentioned in a recent chat, this test as it is doesn't install packages anymore, because it finds the package already present in the `upstream` install tree. 

It would be good to explain such concepts before this paragraph, like what are the different install trees available and how they relate to each other. Maybe show an example using `spack find ...` here rather than in the "test" paragraph below.

Then, the test should be modified to actually install some packages (even just a couple), so it's more realistic and it's actually useful to test the installation procedure.

> To test that your Spack installation works as expected, we will create an `ACCESS-TEST` environment and build the relevant packages (this will take approximately 30 minutes). Then, we will uninstall all the packages and remove the environment.

Reviewer: ```suggestion
To test that _Spack_ works as expected, we will create an `ACCESS-TEST` environment and build the relevant packages (this will take approximately 30 minutes). Then, we will uninstall all the packages and remove the environment.
```

### Find all installed Spack packages

```
spack find
```

### Find all local Spack packages

```
spack find --install-tree local
```

### Clone a Spack environment

> Change directory to the location were you want the model depoloyment repository to reside. e.g. `/g/data/$PROJECT/$USER/`

Reviewer: ```suggestion
Change directory to the location were you want the model deployment repository to reside (e.g., `/g/data/$PROJECT/$USER/`).
```

### Find all installed Spack packages

```
spack find
```

### Find all local Spack packages

```
spack find --install-tree local
```

### Find all upstream Spack packages

```
spack find --install-tree upstream
```

### Clone a Spack environment

Change directory (`cd`) to the location where you want the Model Deployment Repository (MDR) to reside. For example, `/g/data/$PROJECT/$USER/`
```
git clone https://github.com/ACCESS-NRI/ACCESS-TEST.git
```

!!! tip
    If you are creating the _Spack_ environment using `spack create`, we recommend using [_Spack_ _independent_ environments](https://spack.readthedocs.io/en/latest/environments.html#independent-environments). Using independent environments, each clone of a model deployment repository carries its own _Spack_ environment specification (`spack.yaml` file), rather than relying on a [centrally managed environment](https://spack.readthedocs.io/en/latest/environments.html#creating-a-managed-environment).

> 

Author: * We decided not to refer users to https://spack.readthedocs.io/en/latest/environments.html . It's structure is confusing.
* `spack create` creates an SPR not an environment.
* We do not want regular users creating environments.

Reviewer: I agree this admonition can be removed.

!!! warning
    **01/05/2026:** ACCESS-NRI has migrated to a shared _Spack_ `v1.1` instance. If you previously followed these instructions to set up a personal _Spack_ instance, please consider [moving the old _Spack_ directories](#move-old-spack-directories) before using the shared _Spack_ instance.

# How to use Spack on Gadi for building ACCESS models

[Spack](https://spack.io/about/) is a build-from-source package manager, specifically designed to simplify the installation of scientific software on supercomputers. _Spack_ allows users to build ACCESS models directly from the source code, swap model components, and carry out development testing that involves modifying the source code.

> To use _Spack_, please familiarise yourself with the [Basic Usage instructions](https://spack.readthedocs.io/en/latest/basic_usage.html) or follow the instructions in the [Test Spack](#test-spack-optional) section.

Reviewer 2: I think this Test Spack link is broken now? (Maybe that section got re-named at some point?)

Author: Thanks, I appended the expected time to complete the steps to the heading: `Test Spack (OPTIONAL) - 40 mins`
Let's move the time out of the heading and come up with a location that can be adopted for all Hive-Docs.

For instructions on how to build an ACCESS model using _Spack_, refer to [Modify and build an ACCESS model's source code](/models/build_a_model/build_source_code).

## Prerequisites
- **NCI Account**<br> 
    These instructions are tailored specifically for _Gadi_. To use _Spack_ on _Gadi_, you need to [Set Up your NCI Account](/getting_started/set_up_nci_account).

## Enable Spack - 1 minute

!!! warning

>     The ACCESS-NRI shared Spack instance is configured to use `/g/data/$PROJECT/$USER/spack/$SPACK_VERSION` for temporary and permanent files. If your default project id changes, then Spack will not be using the temporary and permenant files created with your previous project id.

Reviewer 2: I'm not sure why this is important? At a guess, I'm assuming it will put things like my build files in there so if my project id changed then I would have to re-make all those things?

Author: > if my project id changed then I would have to re-make all those things?

One issue is that you will have to reinstall Spack packages. The other issue is that the previously installed Spack packages will disappear. That could be very confusing for a user.

Reviewer 2: Ok, my thinking is that if it's possible to be explicit then it helps users diagnose their own problems when issues arise? Still, if there's multiple potential issues then maybe it's impractical. (Related: I don't see an obvious step where one installs Spack packages in this process.)

```
module use /g/data/vk83/modules
module load spack
```

!!! tip
    Please refer to the [ACCESS-NRI Spack Cheat Sheet](https://forum.access-hive.org.au/t/access-nri-spack-cheat-sheet/5942) for help.

> 

Reviewer 2: I wonder if at this point it's worth having this sentence [1]? Maybe even moving it from it's current location in the header. Or perhaps a shorter version of it?

It's like, okay, I've loaded Spack, now what?

[1]
> For instructions on how to build an ACCESS model using _Spack_, refer to [Modify and build an ACCESS model's source code](/models/build_a_model/build_source_code).

Related:  is testing Spack really that much faster than just building a model?

Author: `ACCESS-TEST` was specifically created to be quicker than building a coupled model. It will become even quicker once we define and install an `access-toolchain` in the Spack Release instance.

Reviewer 2: Ah cool, makes more sense :)

!!! warning

>     **18/05/2026:** ACCESS-NRI has migrated to a shared _Spack_ `v1.1` instance. If you previously followed these instructions to set up a personal _Spack_ instance, please read the instructions below.

Reviewer: Not sure how much the "shared spack v1.1 instance" information is needed here. 
What I think is important to convey, is the recent changes in how spack should be enabled on Gadi.
Therefore, I would change to:

```suggestion
!!! warning
    If you previously set up a personal _Spack_ instance by following the instructions on this page, please note that the recommended method for enabling _Spack_ on _Gadi_ has changed. Review the updated instructions below before proceeding.
```

Author: Thanks, I want to keep the date so that there's a clear reference to when this changed.

# How to use Spack on Gadi for building ACCESS models

[Spack](https://spack.io/about/) is a build-from-source package manager, specifically designed to simplify the installation of scientific software on supercomputers. _Spack_ allows users to build ACCESS models directly from the source code, swap model components, and carry out development testing that involves modifying the source code.

> To use _Spack_, please familiarise yourself with the [essential Spack commands](https://spack.readthedocs.io/en/latest/package_fundamentals.html) used for _Listing Available Packages_ and _Seeing Installed Packages_. Alternatively, follow the instructions in the [Test Spack](#test-spack-optional) section.

Reviewer: ```suggestion
To use _Spack_, please familiarise yourself with the [essential Spack commands](https://spack.readthedocs.io/en/latest/package_fundamentals.html) used for [Listing Available Packages](https://spack.readthedocs.io/en/latest/package_fundamentals.html#listing-available-packages) and [Seeing Installed Packages](https://spack.readthedocs.io/en/latest/package_fundamentals.html#seeing-installed-packages). Alternatively, follow the instructions in the [Test Spack](#test-spack-optional) section below.
```

Author: Happy to make this change. Hopefully, the section headings don't change.

## Prerequisites
- **NCI Account**<br> 
    These instructions are tailored specifically for _Gadi_. To use _Spack_ on _Gadi_, you need to [Set Up your NCI Account](/getting_started/set_up_nci_account).

## Enable Spack

Estimated time to complete: 1 minute

!!! warning

>     The ACCESS-NRI shared Spack instance is configured to use `/g/data/$PROJECT/$USER/spack/$SPACK_VERSION` for temporary and permanent files. If your default project id changes, then Spack will not be using the temporary and permanent files created with your previous project id.

Reviewer: ```suggestion
    The ACCESS-NRI shared Spack instance is configured to use `/g/data/$PROJECT/$USER/spack/$SPACK_VERSION` for temporary and permanent files. If your default project ID changes, then _Spack_ will not be using the temporary and permanent files created with your previous project ID. This is particularly relevant for _Spack_ workflows run through PBS jobs using multiple projects.
```

I added a mention to PBS jobs because although this is relevant also for login nodes if users temporarily (e.g., `switchproj ...`) or permanently (e.g., changing their `~/.config/gadi-login.conf` file) change their `$PROJECT`, that doesn't happen often.
What might happen more frequently though, is they run PBS jobs under different projects, and if these PBS jobs use _Spack_, its configured directories will change based on the project the PBS job runs under.

```
module use /g/data/vk83/modules
module load spack
```

!!! tip

>     Please refer to the [ACCESS-NRI Spack Cheat Sheet](https://forum.access-hive.org.au/t/access-nri-spack-cheat-sheet/5942) for help.

Reviewer: ```suggestion
    Please refer to the [ACCESS-NRI Spack Cheat Sheet](https://forum.access-hive.org.au/t/access-nri-spack-cheat-sheet/5942) for further help on _Spack_ commands.
```

Author: Changing to "for further help on _Spack_."

## Test Spack (OPTIONAL)

<!--
Wed May 13 14:18:53 AEST 2026
- setup environment
- concretize
Wed May 13 14:25:55 AEST 2026
- install
Wed May 13 14:57:32 AEST 2026
- cleanup
Wed May 13 14:58:55 AEST 2026
-->

<!-- When the access-toolchain is defined and installed, we need to update the time it takes because it should be quicker. -->

Estimated time to complete: 40 minutes

> To test that _Spack_ works as expected, we will clone the `ACCESS-TEST` repository and use it as a Spack `independent environment` to install the relevant packages. If the packages are already installed in the upstream cache, they will not be rebuilt. Then, we will uninstall all the newly installed packages and remove the environment.

Reviewer: I would not include the cleanup steps in this explanation. They are not the main focus of this test.

```suggestion
To test that _Spack_ works as expected, we will clone the `ACCESS-TEST` repository and use it as a _Spack_ `independent environment` to install the relevant packages. If the packages are already installed in the upstream cache they will not be rebuilt.
```

Also:

> we will clone the `ACCESS-TEST` repository and use it as a _Spack_ `independent environment` to install the relevant packages.

isn't the Spack environment specification the `spack.yaml` file within the ACCESS-TEST repo? In that sense I'm not sure we should say we use ACCESS-TEST as a Spack independent environment.

I would say:
```suggestion
To test that _Spack_ works as expected, we will clone the `ACCESS-TEST` repository and use its `spack.yaml` environment specification to activate `access-test` _Spack_ environment and build its relevant packages. If the packages are already installed in the _Spack_ upstream cache they will not be rebuilt.
```

Author: The Spack environment is the directory that contains the `spack.yaml`:
```
[root@68d8800cf9d7 example]# spack env create ex1
==> Created environment ex1 in: /opt/environments/ex1
==> Activate with: spack env activate ex1

[root@68d8800cf9d7 example]# spack env create ./ex2
==> Created independent environment in: /opt/example/ex2
==> Activate with: spack env activate ./ex2

[root@68d8800cf9d7 example]# file /opt/environments/ex1 /opt/example/ex2
/opt/environments/ex1: directory
/opt/example/ex2:      directory
```

Author: Changed to:
"To test that _Spack_ works as expected, clone the `ACCESS-TEST` repository and use it as a _Spack_ `independent environment` to install the relevant packages. The _Spack_ environment rules are defined in the `ACCESS-TEST/spack.yaml` file. If the packages are already installed in an [upstream](https://spack.readthedocs.io/en/latest/chain.html) they will not be rebuilt."

> ### Find all installed Spack packages
> 
> ```
> spack find
> ```
> 
> ### Find all local Spack packages
> 
> ```
> spack find --install-tree local
> ```
> 
> ### Find all upstream Spack packages
> 
> ```
> spack find --install-tree upstream
> ```

Reviewer: ```suggestion
### Find installed Spack packages

#### All Spack packages

```
spack find
```

#### Local Spack packages

```
spack find --install-tree local
```

#### Upstream Spack packages

```
spack find --install-tree upstream
```
```

Reviewer: We don't mention what "upstream" means here. 

I think we should write a paragraph before the "Test Spack" one, explaining more concepts like "shared" instance, "local", "upstream", etc.

Author: I've added in the preceding paragraph:
"If the packages are already installed in an [upstream](https://spack.readthedocs.io/en/latest/chain.html) they will not be rebuilt."

Author: I've also added a new paragraph in the introduction:
"ACCESS-NRI has installed and configured a shared _Spack_ instance on Gadi so that users do *not* need to create and maintain their own _Spack_ instance."

### Clone a Spack environment

> Change directory (`cd`) to the location where you want the Model Deployment Repository (MDR) to reside. For example, `/g/data/$PROJECT/$USER/`

Reviewer: ```suggestion
Change directory (`cd`) to the location where you want the ACCESS-TEST Repository to reside. For example, `/g/data/$PROJECT/$USER/`
```

If the example is based on the ACCESS-TEST repository, we should be specific about it. 

If you think is useful, we can mention that the ACCESS-TEST repository can be treated as an example of a MDR at the beginning of the Test Spack paragraph, when we illustrate what we're going to do.

    <terminal-line data="input" directory="[ACCESS-TEST]" class="spack" lineDelay=0></terminal-line>
</terminal-window>

!!! tip

>     The command above is creating and activating a [ _Spack_ independent environment](https://spack.readthedocs.io/en/latest/environments.html#independent-environments). In this example, we are not creating the environment using `spack create` because the `spack.yaml` file already exists in the `ACCESS-TEST` folder. If you wish to modify the specifics of the build for `ACCESS-TEST`, then you can directly edit the `spack.yaml` file before activating the environment using the command above.

Author: * We decided not to refer users to https://spack.readthedocs.io/en/latest/environments.html . It's structure is confusing.
* `spack create` creates an SPR not an environment.
* We do not want regular users creating environments.

Reviewer: I'm fine with removing this.

rm -rf ACCESS-TEST
```

## Develop a Model

For instructions on how to build an ACCESS model using _Spack_, refer to [Modify and build an ACCESS model's source code](/models/build_a_model/build_source_code).

## Advanced users

> For those who want to use Spack outside of Gadi, we also recommend that you refer to the [Spack 101 Tutorial](https://spack-tutorial.readthedocs.io/en/latest/).

Reviewer: ```suggestion
For those who want to use _Spack_ outside of _Gadi_, we also recommend that you refer to the [Spack 101 Tutorial](https://spack-tutorial.readthedocs.io/en/latest/).
```

> [ACCESS-NRI's Spack configuration](https://github.com/ACCESS-NRI/spack-config) contains the details on how ACCESS-NRI's Spack configuration can be installed on another HPC.

Reviewer: ```suggestion
[ACCESS-NRI's Spack configuration](https://github.com/ACCESS-NRI/spack-config) contains the details on how ACCESS-NRI's _Spack_ configuration can be installed on another HPC.
```
